### PR TITLE
[WIP] Public partition selection

### PR DESCRIPTION
**Pull request by lovroprepolec**
## Description
This is early implementation of https://github.com/OpenMined/PipelineDP/issues/4 for Apache beam. Sharing this early as I'm OOO until 2021-06-01 to get feedback.

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
Unit tests.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: tests/pipeline_operations_test.py [hunk 1]
```diff
@@ -1,14 +1,39 @@
...
+          value_extractor=lambda x: x[2])
+
+    def test_filter_partitions_noop(self):
+      col = [(1, 6, 1), (2, 7, 1), (3, 6, 1), (4, 7, 1), (5, 8, 1)]
+      public_partitions = []
+      result = self.ops.filter_partitions(col, public_partitions, self.data_extractors, "Public partition filtering")
+      assert result == col
+
+
+    def test_filter_partitions_remove(self):
```
Comment (dvadym) [inline tests/pipeline_operations_test.py:30]: could you please also when public_partition is PCollection?

Here is the how to test Beam pipeline https://beam.apache.org/documentation/pipelines/test-your-pipeline/

File: pipeline_dp/pipeline_operations.py [hunk 2]
```diff
@@ -77,6 +81,48 @@ def group_by_key(self, col, stage_name: str):
...
+        # no filtering if no public partitions are specified
+        if not public_partitions:
+          return col
+
+        col = col | beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        pp_type = type(public_partitions)
+        if pp_type is list or pp_type is set:
+          return col | beam.Filter(is_public)
+        else:
```
Comment (dvadym) [inline pipeline_dp/pipeline_operations.py:122]: Great for implementing both in-memory and join based partition selection!

**Review by dvadym — CHANGES_REQUESTED** [review]
Thanks! I left some comments

If needed some help with Beam, please LMK, we can have a pair programming session.

File: examples/movie_view_ratings.py [hunk 3]
```diff
@@ -83,6 +83,7 @@ def calc_dp_rating_metrics(pipeline):
       max_contributions_per_partition=1,
       low=1,
       high=5,
+      public_partitions=[int(partition) for partition in FLAGS.public_partitions]
```
[the remark this answers is not in the document]
Comment (lovroprepolec) [reply examples/movie_view_ratings.py:86]: Done.

File: pipeline_dp/pipeline_operations.py [hunk 4]
```diff
@@ -77,6 +81,48 @@ def group_by_key(self, col, stage_name: str):
     def filter(self, col, fn, stage_name: str):
         return col | stage_name >> beam.Filter(fn)
 
+    def filter_partitions(self, col, public_partitions, data_extractors, stage_name: str):
```
[the remark this answers is not in the document]
Comment (lovroprepolec) [reply pipeline_dp/pipeline_operations.py:84]: Done.

File: pipeline_dp/pipeline_operations.py [hunk 5]
```diff
@@ -77,6 +81,48 @@ def group_by_key(self, col, stage_name: str):
...
+          return col
+
+        col = col | beam.Map(lambda x: (data_extractors.partition_extractor(x), x))
+
+        pp_type = type(public_partitions)
+        if pp_type is list or pp_type is set:
+          return col | beam.Filter(is_public)
+        else:
+          public_partitions = public_partitions | beam.Map(lambda x: (x, True))
+          return ({'values': col, 'is_public': public_partitions} | beam.CoGroupByKey() | beam.ParDo(PartitionsFilterJoin()))
```
[the remark this answers is not in the document]
Comment (lovroprepolec) [reply pipeline_dp/pipeline_operations.py:124]: Sure! Thanks.

File: pipeline_dp/pipeline_operations.py (hunk 2 above)
Comment (lovroprepolec) [reply pipeline_dp/pipeline_operations.py:122]: Thanks!

File: pipeline_dp/pipeline_operations.py [hunk 6]
```diff
@@ -77,6 +81,48 @@ def group_by_key(self, col, stage_name: str):
...
+
+            if is_public:
+              for value in values:
+                yield key, value
+
+        def is_public(col):
+          return col[0] in public_partitions
+
+        # no filtering if no public partitions are specified
+        if not public_partitions:
```
[the remark this answers is not in the document]
Comment (lovroprepolec) [reply pipeline_dp/pipeline_operations.py:114]: True. However, then it gets difficult to test noop in the pipeline_operations_test.py. Any suggestions?

[the remark this answers is not in the document]
Comment (dvadym) [reply pipeline_dp/pipeline_operations.py]: Yeah, good question, what should be behaviour of filter_by_key. 

My feeling that filter_by_key shouldn't be called with None public_partitions. Probably raising exception is a good option. WDYT?

[the remark this answers is not in the document]
Comment (lovroprepolec) [reply pipeline_dp/pipeline_operations.py]: SG. Thx.

**Pull request merged**